neutromeratio/restraints.py

```diff
@@ -1,122 +1,194 @@
 import numpy as np
 import torch
-from .constants import bond_length_dict
 import logging
 from scipy.stats import norm
 from torch.distributions.normal import Normal
 from openmmtools.constants import kB
+import torch
 
 logger = logging.getLogger(__name__)
 
 
-def flat_bottom_position_restraint(x, tautomer_transformation:dict, atom_list:list, restrain_acceptor:bool, restrain_donor:bool, device:torch.device):
-    
-    """
-    Applies a flat bottom positional restraint.
-
-    Parameters
-    ----------
-    x0 : array of floats, unit'd (distance unit)
-        initial configuration
-    tautomer_transformation : dict
-        dictionary with index of acceptor, donor and hydrogen idx
-    atom_list : list 
-        list of elements
-    restrain_acceptor : bool
-        should the acceptor be restraint
-    restrain_donor : boold
-        should the donor be restraint
-
-    Returns
-    -------
-    e : float
-        bias
-    """
-
-    k = 5
-    if restrain_acceptor:
-        heavy_atom_idx = tautomer_transformation['acceptor_idx']
-        if 'acceptor_hydrogen_idx' in tautomer_transformation:
-            hydrogen_idx = tautomer_transformation['acceptor_hydrogen_idx']
-        else:
-            hydrogen_idx = tautomer_transformation['hydrogen_idx']
-    elif restrain_donor:
-        heavy_atom_idx = tautomer_transformation['donor_idx']
-        if 'donor_hydrogen_idx' in tautomer_transformation:
-            hydrogen_idx = tautomer_transformation['donor_hydrogen_idx']
+class Restraint(object):
+
+    def __init__(self, sigma:unit.Quantity, atom_i_idx:int, atom_j_idx:int, atoms:str, active_at_lambda:int):
+        """
+        Defines a restraint. 
+
+        Parameters
+        ----------
+        sigma : in angstrom
+        atom_i_idx : int
+            Atom i to restraint
+        atom_j_idx : int
+            Atom j to restraint
+        atoms: str
+            Str of atoms to retrieve element information
+        active_at_lambda : int
+            Integer to indicccate at which state the restraint is fully active. Either 0 (for 
+            lambda 0), or 1 (for lambda 1) or -1 (always active)
+        Returns
+        -------
+        e : float
+            bias
+        """
+
+        assert(type(sigma) == unit.Quantity)
+        k = (kB * temperature) / (sigma**2)
+        self.k = k.value_in_unit((unit.kilo * unit.joule) / ((unit.angstrom **2) * unit.mole))
+        self.device = device
+        self.atom_i_element = atoms[atom_i_idx]
+        self.atom_j_element = atoms[atom_j_idx]
+        self.atom_i_idx = atom_i_idx
+        self.atom_j_idx = atom_j_idx
+        self.mean_bond_length = (bond_length_dict[frozenset([self.atom_i_element, self.atom_j_element])]).value_in_unit(unit.angstrom)
+        self.active_at_lambda = active_at_lambda
+        self.upper_bound = self.mean_bond_length + 0.2
+        self.lower_bound = self.mean_bond_length - 0.2
+
+
+    def flat_bottom_position_restraint(self, x):
```

Comment:
instead of having two functions in the same class, could we have two subclasses that inherit from an abstract `Restraint` class? 

`Restraint(object)` (with instance method `.energy(x)`): subclasses`FlatBottomRestraint(Restraint)`, `HarmonicRestraint(Restraint)` ...

Reply:
good point! will put on the TODO list!